Задание 2: API

Anna-thorn opened the pull request:
Создание пользователя:
создать уникального пользователя;
создать пользователя, который уже зарегистрирован; создать пользователя и не заполнить одно из обязательных полей.

Логин пользователя:
логин под существующим пользователем,
логин с неверным логином и паролем.

Изменение данных пользователя:
с авторизацией,
без авторизации,
Для обеих ситуаций нужно проверить, что любое поле можно изменить. Для неавторизованного пользователя — ещё и то, что система вернёт ошибку.

Создание заказа:
с авторизацией,
без авторизации,
с ингредиентами,
без ингредиентов,
с неверным хешем ингредиентов.

Получение заказов конкретного пользователя:
авторизованный пользователь,
неавторизованный пользователь.

Commit:
Задание 2: API
Создание пользователя:
создать уникального пользователя;
создать пользователя, который уже зарегистрирован;
создать пользователя и не заполнить одно из обязательных полей.

Логин пользователя:
логин под существующим пользователем,
логин с неверным логином и паролем.

Изменение данных пользователя:
с авторизацией,
без авторизации,
Для обеих ситуаций нужно проверить, что любое поле можно изменить. Для неавторизованного пользователя — ещё и то, что система вернёт ошибку.

Создание заказа:
с авторизацией,
без авторизации,
с ингредиентами,
без ингредиентов,
с неверным хешем ингредиентов.

Получение заказов конкретного пользователя:
авторизованный пользователь,
неавторизованный пользователь.

Commit:
Добавлен pom.xml

Commit:
add allure report

okryukova commented on pom.xml:
⚠️Можно улучшить. Хорошей практикой является добавление README.md с описанием проекта, которое включает название, технологии в проекте (например, Java 11, JUnit 4.13.2, maven 3.9.0 и т.д.), настройку (если необходима) и как запускать (mvn clean test - кавычки позволяют выполнять запуск из файла README.md). Для оформления данных в виде таблицы используй сочетание клавиш Ctrl + N или Cmd + N. Как оформлять README.md можно ознакомиться в статье https://blog.jetbrains.com/idea/2022/08/markdown-support-improvements/

okryukova commented on src/main/java/api/OrderApi.java:
```diff
@@ -0,0 +1,48 @@
+ * Класс с методами API для заказов
+ */
+public class OrderApi {
+    private static final String BASE_URL = "https://stellarburgers.nomoreparties.site";
```
⛔️Нужно исправить.  Base URL присваиваем один раз в проекте. Можно использовать request specification в RestAssured или утилитный базовый класс.

okryukova commented on src/main/java/models/User.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Модель пользователя для API тестов.
+ */
+public class User {
```
⚠️Можно улучшить. Можно использовать @DaTa и @AllArgsConstructor из библиотеки Lombok

okryukova commented on src/main/java/models/User.java:
```diff
@@ -0,0 +1,37 @@
+    public void setName(String name) {this.name = name;}
+
+    public static User getRandomUser() {
+        String randomString = UUID.randomUUID().toString().substring(0, 8);
```
⚠️Можно улучшить. Лучше использовать готовые решения JavaFaker/DataFaker, так как это сокращает нашу работу.

okryukova commented on src/test/java/OrderApiTest.java:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    @Test
+    @DisplayName("Создание заказа без авторизации с ингредиентами")
```
⚠️Можно улучшить. В allure есть еще одна аннотация для описания теста Description. Добавь ее, пожалуйста, во всех тестах проекта

okryukova commented on src/test/java/UserCreationApiTest.java:
```diff
@@ -0,0 +1,99 @@
+        assertNotNull(json.getString("refreshToken"));
+    }
+
+    private void verifyErrorResponse(Response response, String expectedMessage) {
```
⛔️Нужно исправить. Для всех методов, описывающих шаги теста, необходимо использовать аннотацию @Step

Commit:
Исправления:
1. Добавлен BaseApi и внесены соответствующие изменения в пакете api
2. В pom.xml добавлены зависимостии: Project Lombok и Datafaker
3. В User
     - добавлены @Data, @AllArgsConstructor и @NoArgsConstructor
     - использован DataFaker
     - удален лишний код
4. В Order добавлены @Data, @AllArgsConstructor
5. В OrderApiTest добавлена к тестам @Description
6. В UserCreationApiTest добавлен @Step и @Description
7. В UserLoginApiTest добавлена к тестам @Description
8. В UserUpdateApiTest добавлена к тестам @Description

Commit:
add allure report2

okryukova commented on src/test/java/UserUpdateApiTest.java:
```diff
@@ -0,0 +1,134 @@
+        String newPassword = randomUser.getPassword();
+        User updateData = new User(null, newPassword, null);
+        Response updateResponse = userApi.updateUserData(accessToken, updateData);
+        assertEquals("Неверный код ответа при обновлении пароля",
```
⛔️Нужно исправить. Для тестов придерживаемся подхода: один тест, значит одна проверка. Если очень хочется несколько проверок -- тогда используем softAssertions. Поправь, пожалуйста, во всем коде

Commit:
Исправления:
1. В pom.xml добавлена зависимость AssertJ (для SoftAssertions)
2. В тестах добавлено использование softAssertions

Commit:
Исправлен createOrderWithInvalidIngredientHash

Commit:
add allure report3
